src/util/key_reader.py

```diff
@@ -0,0 +1,99 @@
+from config import LIST_ARDUINO_SERIAL_DEVICES_PATH
+from src.util.logger import get_logger
+import subprocess
+import serial
+import re
+
+logger = get_logger()
+
+class NoReaderFound(Exception):
+    pass
+
+def abstract(fun):
+    def abstract_wrapper(*args, **kwargs):
+        raise NotImplementedError("This is only and abstract function")
+    return abstract_wrapper
+
+class KeyReader(object):
+    @classmethod
+    @abstract
+    def get_devices(cls):
+        pass
+
+    @abstract
+    def tag_was_read(self):
+        return False
+
+    @classmethod
+    def get_reader(cls):
+        key_readers = cls.get_devices()
+        if len(key_readers) == 0:
+            raise NoReaderFound()
+
+        key_reader = key_readers[0]
+        if len(key_readers) > 1:
+            logger.warning(f"There are several key readers connected: {key_readers}. Choose {key_reader}")
+
+        return key_reader
+
+class EM4100_KeyReader(KeyReader):
+    def __init__(self, serial_device):
+        self.serial_device = serial_device
+        logger.info(f"Connecting to serial port {serial_device}")
+
+        # Arduino is restarted when serial port is opened
+        self.com = serial.Serial(port=serial_device, baudrate=115200, timeout=2)
+        com = self.com
+        com.readline() # Just wait until it starts up and starts printing something (hopefully less than 2 second timeout)
+        com.timeout = 0
+        self.check_echo()
+        self.last_tag_id = None
+
+    def __repr__(self):
+        return f"<EM4100 Key Reader tty={self.serial_device}>"
+
+    def tag_was_read(self):
+        if self.com.in_waiting == 0:
+            return False
+        lines = self.com.read(10000).decode("utf-8").split("\r\n")
```

Comment:
10000 is arbitrarily chosen. We can only get 115200 byte per second at most from the serial port.

Reply:
Replacing magic constant with actual number of bytes in pipe